funsor/registry.py

```diff
@@ -50,8 +51,11 @@ def register(self, key, *types):
         register = self.registry[key].register
         if self.default:
             objects = (object,) * len(types)
-            if objects != types:
-                register(*objects)(self.default)
+            try:
+                if objects != types and supercedes(types, objects):
+                    register(*objects)(self.default)
+            except TypeError:
+                pass  # mysterious source of ambiguity in Python 3.5 breaks this
```

Comment:
This is the only significant change in this PR.  It places stricter constraints on method resolution order in multiple dispatch.  If this is reverted, attempting to run the test suite in Python 3.5 will lead to a bunch of strange, nondeterministic and difficult-to-diagnose test failures ultimately caused by patterns not firing when they should.

Reply:
Weird. Do you know whether the Type error occurs in `!=`, `supercedes()`, `and`, or `register()()`?

Reply:
Adding `supercedes` here fixes the test failures, but introduces occasional `TypeError`s which can be disregarded.